src/sbvr-api/control-flow.ts

```diff
@@ -0,0 +1,61 @@
+import * as _ from 'lodash'
+import * as Promise from 'bluebird'
+
+// lifts f to map over the changeSet if present
+export const liftP = <T, U>(fn: (v: T) => U | Promise<U>) => {
+	return ((a: T | T[]) => {
+		if (_.isArray(a)) {
+			// This must not be a settle as if any operation fails in a cs
+			// we want to discard the whole
+			return Promise.mapSeries(a, fn)
+		} else {
+			return Promise.resolve(a).then(fn)
+		}
+	// TODO: We cast here to merge `Promise<U> | Promise<U[]>` into `Promise<U | U[]>,
+	// which is equivalent for our uses but fixes issues where typing was being rejected
+	}) as ((a: T | T[]) => Promise<U | U[]>)
+}
+
+interface MappingFunction {
+	<T, U>(a: T[], fn: (v: T) => U | U[] | Promise<U>): Promise<Array<U | U[] | Error>>
+}
+
+// The settle- versions of
+const settleMapSeries: MappingFunction = (a, fn) => {
+	const runF = Promise.method(fn)
+	return Promise.mapSeries(a, _.flow(runF, wrap))
+}
+
+// Wrap a promise with reflection. This promise will always succeed, either
+// with the value or the error of the promise it is wrapping
+const wrap = <T>(p: Promise<T>) => {
+	return p.then(_.identity as (value: T) => T, _.identity as (value: Error) => Error)
+}
```

Comment:
Minor: I like this a bit better as 
```
const wrap = <T>(p: Promise<T>):(Promise<T|Error>) => {
	return p.then(_.identity, _.identity)
}
```

Reply:
If I try that I get:
```
Type 'Bluebird<undefined>' is not assignable to type 'Bluebird<Error | T>'.
  Type 'undefined' is not assignable to type 'Error | T'.

Reply:
Hmmm this is also apparently due to a patch version mismatch. I had @types/lodash@4.14.64 and it works fine with that, updating to 4.14.77 gives me the same error you have. (Which might be a regression as the error doesn't really make much sense)

Reply:
I've updated package.json to the latest versions to help a bit with with this